File: src/Scripting/Ruby/patterns/enterprise_facade.rb
# frozen_string_literal: true

module RubyPatterns
  module EnterpriseFacade
    module_function

    def run
      stock = -> { true }
      charge = -> { :paid }
      checkout = -> { stock.call ? charge.call : :sold_out }
      raise 'Enterprise Facade failed' unless checkout.call == :paid
      true
    end
  end
end

RubyPatterns::EnterpriseFacade.run if $PROGRAM_NAME == __FILE__

File: src/Scripting/Lua/patterns/enterprise_facade.lua
local inventory = function() return true end
local payment = function() return "paid" end
local shipping = function() return "tracking-1" end
local facade = {}
function facade.checkout()
  assert(inventory())
  return payment() .. ":" .. shipping()
end
assert(facade.checkout() == "paid:tracking-1")
return true
